caravel/assets/javascripts/explorev2/actions/exploreActions.js

```diff
@@ -19,89 +14,52 @@ export const CLEAR_ALL_OPTS = 'CLEAR_ALL_OPTS';
-  return { type: SET_ORDERING_OPTS, orderingOpts };
+export const FETCH_STARTED = 'FETCH_STARTED';
+export function fetchStarted() {
+  return { type: FETCH_STARTED };
 }
 
-export function setFilterColumnOpts(filterColumnOpts) {
-  return { type: SET_FILTER_COLUMN_OPTS, filterColumnOpts };
+export const FETCH_SUCCEEDED = 'FETCH_SUCCEEDED';
+export function fetchSucceeded() {
+  return { type: FETCH_SUCCEEDED };
 }
 
-export function clearAllOpts() {
-  return { type: CLEAR_ALL_OPTS };
+export const FETCH_FAILED = 'FETCH_FAILED';
+export function fetchFailed() {
+  return { type: FETCH_FAILED };
 }
 
-export function setFormOpts(datasourceId, datasourceType) {
+export function fetchFieldOptions(datasourceId, datasourceType) {
   return function (dispatch) {
-    const timeColumnOpts = [];
-    const groupByColumnOpts = [];
-    const metricsOpts = [];
-    const filterColumnOpts = [];
-    const timeGrainOpts = [];
-    const columnOpts = [];
-    const orderingOpts = [];
+    dispatch(fetchStarted());
 
     if (datasourceId) {
       const params = [`datasource_id=${datasourceId}`, `datasource_type=${datasourceType}`];
       const url = '/caravel/fetch_datasource_metadata?' + params.join('&');
 
       $.get(url, (data, status) => {
         if (status === 'success') {
-          data.time_columns.forEach((d) => {
-            if (d) timeColumnOpts.push({ value: d, label: d });
-          });
-          data.groupby_cols.forEach((d) => {
-            if (d) groupByColumnOpts.push({ value: d, label: d });
-          });
-          data.metrics.forEach((d) => {
-            if (d) metricsOpts.push({ value: d[1], label: d[0] });
-          });
-          data.filter_cols.forEach((d) => {
-            if (d) filterColumnOpts.push({ value: d, label: d });
-          });
-          data.time_grains.forEach((d) => {
-            if (d) timeGrainOpts.push({ value: d, label: d });
-          });
-          data.columns.forEach((d) => {
-            if (d) columnOpts.push({ value: d, label: d });
-          });
-          data.ordering_cols.forEach((d) => {
-            if (d) orderingOpts.push({ value: d, label: d });
-          });
-
-          // Repopulate options for controls
-          dispatch(setTimeColumnOpts(timeColumnOpts));
-          dispatch(setTimeGrainOpts(timeGrainOpts));
-          dispatch(setGroupByColumnOpts(groupByColumnOpts));
-          dispatch(setMetricsOpts(metricsOpts));
-          dispatch(setFilterColumnOpts(filterColumnOpts));
-          dispatch(setColumnOpts(columnOpts));
-          dispatch(setOrderingOpts(orderingOpts));
+          // populate options for select type fields
+          dispatch(setFieldOptions(data.field_options));
+          dispatch(fetchSucceeded());
+        } else if (status === 'error') {
+          dispatch(fetchFailed());
         }
       });
     } else {
-      // Clear all Select options
-      dispatch(clearAllOpts());
+      // in what case don't we have a datasource id?
```

Comment:
i'm not sure it's needed either. in the explore view we should always have a datasource. we can remove in a follow up PR.